research/2022-01-17-speeding-up-evm-part-1.mdx

```diff
@@ -0,0 +1,155 @@
+- ``{(r), (r)}``: $txn_i$ and $txn_j$ are parallelizable, assuming that $\sigma$ is only one in the "read" set of both those transactions.
+- ``{(r), (w)}``: $txn_i$ and $txn_j$ must be executed sequentially in order of $\bar{t}$.
+- ``{(w), (r)}``: $txn_i$ and $txn_j$ must be executed sequentially in order of $\bar{t}$.
+- ``{(w), (w)}``: if the write operation to $s'$ is [commutative](https://arxiv.org/abs/1802.08748) (e.g. addition on unsigned integers is a commutative operation), then $txn_i$ and $txn_j$ are parallelizable, or else they must be executed in order of $\bar{t}$.[^9]
+
+[^9]: Examples of (non)commutative writes in smart contracts can be found [here](https://dl.acm.org/doi/pdf/10.1145/3453483.3454112).
+
+However, $txn_i$ and $txn_j$ access not only $\sigma$ but more locations, so we extend our four simple rules to include a read set and a write set for every transaction, and when searching for parallelizable transactions to execute, we loop through every transactions' storage access information $\kappa$ and apply the above rules.
+
+Alternatively, we could use the simple algorithm described by Vitalik in [EIP 648](https://github.com/ethereum/EIPs/issues/648): each transaction includes a set $\beta$ of the addresses it accesses, and if two transactions $txn_i$ and $txn_j$ satisfy $\beta_i \cap \beta_j = \emptyset$, then execute them in parallel, else not.
+
+Ultimately, it all depends on how refined our $\kappa$ is and how refined we want our parallel execution to be.[^10] For example, it could be more than [quadratic](https://www.google.com/search?q=quadratic+sharding), meaning that we have $\kappa$ not only contain storage access information but also those on memory/calldata, as we are parallelizing within a single transaction as well.
+
+[^10]: Other blockchains such as Solana [greedily](https://jito-labs.medium.com/solana-validator-101-transaction-processing-90bcdc271143) (based on validator's $\kappa$, which is the signed account addresses every transaction has in its signature) separate mempool transactions into different batches for parallel execution and use storage locks to prevent any potential data races.
+
+Of course, within those four situations, there are many complexities. For example: ``{(w), (w)}``. In this situation, it is possible that we have $txn_i$ first reads $s'$ and then changes it, but the value assigned to $s'$ always equals that of $txn_j$'s assignment because of how the smart contract was written. So this effectively reduces to the ``{(r), (r)}`` case. Or this could easily go the other way and reduce to ``{(w), (r)}``,  ``{(w), (w)}``, or ``{(r), (w)}``. And even then it could be that the writer somehow doesn't change the storage's value or that the reader doesn't affect state change in the EVM (e.g. reading some value to be 2 while it was originally 1, but the only use for that storage reading was only a `require` checking if the variable is positive).
```

Comment:
nit but this section was a lil hard to follow for me, just a flag for you to review it carefully & potentially make it simpler if you think it's worth it